weave/trace_server/execution_runner/cross_process_trace_server.py

```diff
@@ -0,0 +1,540 @@
+        """
+        Run a model asynchronously.
+
+        Note: Async methods require special handling. Currently runs synchronously.
+        """
+        # TODO: Implement proper async support
+        return self._send_request("run_model", req)
+
+
+def generate_child_process_trace_server_args(
+    trace_server: TraceServerInterface,
+) -> CrossProcessTraceServerArgs:
+    """
+    Start a worker thread with the trace server and return serializable args.
+
+    This function runs in the PARENT process and:
+    1. Creates multiprocessing queues for communication
+    2. Captures current context variables
+    3. Starts a worker thread that runs the given trace server with context
+    4. Returns ONLY the queues (which are serializable)
+
+    The trace server stays in the parent process, only the queues are passed to child.
+
+    Args:
+        trace_server: The trace server instance to run in the worker thread (stays in parent)
+
+    Returns:
+        A dictionary containing the queues to pass to the child process
+    """
+    # Get the multiprocessing context
+    ctx = multiprocessing.get_context()
+
+    # Create multiprocessing queues (these CAN be serialized)
+    request_queue = ctx.Queue()
+    response_queue = ctx.Queue()
+
+    # Capture current context variable values
+    context_values = {}
+    for var in CONTEXT_VARS_TO_PROPAGATE:
+        try:
+            value = var.get()
+            context_values[var] = value
+        except LookupError:
+            # Context var not set in current context, skip it
+            pass
+
+    # Start the worker thread IN THE PARENT PROCESS
+    worker_thread = threading.Thread(
+        target=_trace_server_worker_loop_with_context,
+        args=(request_queue, response_queue, trace_server, context_values),
+        daemon=True,
+    )
+    worker_thread.start()
+
+    # Return ONLY the queues (not the trace server!)
+    return CrossProcessTraceServerArgs(
+        request_queue=request_queue, response_queue=response_queue
+    )
+
+
+def build_child_process_trace_server(
+    args: CrossProcessTraceServerArgs,
+) -> CrossProcessTraceServer:
+    """
+    Build a CrossProcessTraceServer from serializable arguments.
+
+    This function runs in the CHILD process with queues created in the parent.
+
+    Args:
+        args: The arguments containing the communication queues from parent
+
+    Returns:
+        A CrossProcessTraceServer instance that communicates with the parent's worker
+    """
+    return CrossProcessTraceServer(
+        request_queue=args["request_queue"], response_queue=args["response_queue"]
+    )
+
+
+def _trace_server_worker_loop_with_context(
+    request_queue: multiprocessing.Queue,
+    response_queue: multiprocessing.Queue,
+    trace_server: TraceServerInterface,
+    context_values: dict[contextvars.ContextVar, Any],
+) -> None:
+    """
+    Worker loop that handles requests using the provided trace server with context.
```

Comment:
this is non trivial, maybe a bit more description here....

Reply:
added